components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt

```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
+        )
+        return createFailureWithoutRetryResponse()
+    }
+
+    private fun createMoveOnResponse(): StateAndEventProcessor.Response<MembershipAsyncRequestState> =
+        StateAndEventProcessor.Response(
+            updatedState = null,
+            responseEvents = emptyList(),
+            markForDLQ = false,
+        )
+
+    private fun createFailureWithRetryResponse(
+        value: MembershipAsyncRequest,
+        retries: Int,
+    ): StateAndEventProcessor.Response<MembershipAsyncRequestState> {
+        return StateAndEventProcessor.Response(
+            updatedState = MembershipAsyncRequestState(
+                value,
+                RetriableFailure(retries - 1),
+            ),
+            responseEvents = emptyList(),
+            markForDLQ = false,
+        )
+    }
+
+    private fun createSentToMgmResponse(
+        value: MembershipAsyncRequest,
+        state: MembershipAsyncRequestState?,
+    ): StateAndEventProcessor.Response<MembershipAsyncRequestState> {
+        val currentCause = state?.cause
+        val newCause = if (currentCause is SentToMgmWaitingForNetwork) {
+            currentCause
+        } else {
+            SentToMgmWaitingForNetwork(
+                clock.instant().plusSeconds(waitForMgmSeconds),
             )
-            return Outcome.FAILED_CAN_RETRY
         }
+        return StateAndEventProcessor.Response(
+            updatedState = MembershipAsyncRequestState(
+                value,
+                newCause,
+            ),
+            responseEvents = emptyList(),
+            markForDLQ = false,
+        )
+    }
+
+    private fun MembershipAsyncRequestState?.retries(): Int {
+        return (this?.cause as? RetriableFailure)?.numberOfRetriesToGo ?: MAX_RETRIES
+    }
+
+    private fun register(
+        value: MembershipAsyncRequest,
+        state: MembershipAsyncRequestState?,
+    ): StateAndEventProcessor.Response<MembershipAsyncRequestState> {
+        val request = value.request
         val registrationId = try {
             UUID.fromString(request.requestId)
         } catch (e: IllegalArgumentException) {
             logger.warn("Registration ${request.requestId} failed. Invalid request ID.", e)
-            return Outcome.FAILED_CANNOT_RETRY
+            return createFailureWithoutRetryResponse()
+        }
+        val holdingIdentityShortHash = ShortHash.of(request.holdingIdentityId)
+        val holdingIdentity =
+            virtualNodeInfoReadService.getByHoldingIdentityShortHash(holdingIdentityShortHash)?.holdingIdentity
+        if (holdingIdentity == null) {
+            val retries = state.retries()
+            return if (retries <= 0) {
+                logger.warn(
+                    "Registration ${request.requestId} failed." +
+                        " Could not find holding identity associated with ${request.holdingIdentityId} for too long." +
+                        " Will not retry nor persist the state.",
```

Comment:
```suggestion
                    "Registration ${request.requestId} failed." +
                        " Could not find holding identity associated with ${request.holdingIdentityId} for too long." +
                        " Request will not be retried further.",
```
to make it a bit easier for the person that reads the logs and is unaware of internal details.